File: src/main/java/io/disassemble/javanalysis/flow/FlowBlock.java
package io.disassemble.javanalysis.flow;

import io.disassemble.javanalysis.CtMethodNode;
import io.disassemble.javanalysis.insn.CtInsn;
import javassist.bytecode.analysis.ControlFlow;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
public class FlowBlock {

    private List<CtInsn> instructions = new ArrayList<>();

    private final CtMethodNode method;
    private final ControlFlow.Block source;

    private FlowBlock[] exits;
    private FlowBlock[] incoming;

    public FlowBlock(CtMethodNode method, ControlFlow.Block source) {
        this.method = method;
        this.source = source;
    }

    public ControlFlow.Block source() {
        return source;
    }

    public int startIndex() {
        return source.position();
    }

    public int endIndex() {
        return (startIndex() + source.length());
    }

    public List<CtInsn> instructions() {
        return instructions;
    }

    public FlowBlock[] exits() {
        return exits;
    }

    public int[] exitIndices() {
        int[] indices = new int[exits.length];
        for (int i = 0; i < indices.length; i++) {
            indices[i] = exits[i].startIndex();
        }
        return indices;
    }

    public FlowBlock[] incoming() {
        return incoming;
    }

    public int[] incomingIndices() {
        int[] indices = new int[incoming.length];
        for (int i = 0; i < indices.length; i++) {
            indices[i] = incoming[i].startIndex();
        }
        return indices;
    }

    public ControlFlow.Catcher[] catchers() {
        return source.catchers();
    }

    public void init(ControlFlowGraph cfg) {
        if (exits != null) {
            return; // already initialized
        }
        constructInstructions(method.instructions());
        constructExits(cfg);
        constructIncoming(cfg);
    }

    private void constructInstructions(List<CtInsn> instructions) {
        int start = method.normalizeIndex(startIndex());
        int end = method.normalizeIndex(endIndex());
        int offset = (method.hasIndex(endIndex()) ? 0 : 1); // this occurs on the end source
        this.instructions.addAll(instructions.subList(start, end + offset));
    }

    private void constructExits(ControlFlowGraph cfg) {
        exits = new FlowBlock[source.exits()];
        for (int i = 0; i < exits.length; i++) {
            exits[i] = translate(cfg, source.exit(i));
        }
    }

    private void constructIncoming(ControlFlowGraph cfg) {
        incoming = new FlowBlock[source.incomings()];
        for (int i = 0; i < incoming.length; i++) {
            incoming[i] = translate(cfg, source.incoming(i));
        }
    }

    public static FlowBlock translate(ControlFlowGraph cfg, ControlFlow.Block block) {
        return cfg.findBlockByIndex(block.position());
    }
}

File: src/main/java/io/disassemble/javanalysis/flow/FlowBlock.kt
package io.disassemble.javanalysis.flow

import io.disassemble.javanalysis.hasIndex
import io.disassemble.javanalysis.insn.CtInsn
import io.disassemble.javanalysis.instructions
import io.disassemble.javanalysis.normalizeIndex
import javassist.CtMethod
import javassist.bytecode.analysis.ControlFlow
import java.util.*

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
class FlowBlock(val method: CtMethod, val source: ControlFlow.Block) {

    val instructions = ArrayList<CtInsn>()

    private var outgoingEdges: List<FlowBlock?>? = null
    private var incomingEdges: List<FlowBlock?>? = null

    val startIndex: Int
        get() = source.position()

    val endIndex: Int
        get() = startIndex + source.length()

    val exits: List<FlowBlock?>
        get() = outgoingEdges ?: emptyList()

    val exitIndices: IntArray
        get() {
            outgoingEdges?.let { exits ->
                val indices = IntArray(exits.size)
                for (i in indices.indices) {
                    indices[i] = exits[i]?.startIndex ?: -1
                }
                return indices
            }
            return IntArray(0)
        }

    val incoming: List<FlowBlock?>
        get() = incomingEdges ?: emptyList()

    val incomingIndices: IntArray
        get() {
            incomingEdges?.let { incoming ->
                val indices = IntArray(incoming.size)
                for (i in indices.indices) {
                    // TODO: should -1 be the default?
                    indices[i] = incoming[i]?.startIndex ?: -1
                }
                return indices
            }
            // TODO: should this throw an exception?
            return IntArray(0)
        }

    val catchers: List<ControlFlow.Catcher>
        get() = source.catchers().toList()

    fun init(cfg: ControlFlowGraph) {
        if (outgoingEdges != null) {
            return  // already initialized
        }
        constructInstructions(method.instructions)
        outgoingEdges = constructExits(cfg)
        constructIncoming(cfg)
    }

    private fun constructInstructions(instructions: List<CtInsn>) {
        val start = method.normalizeIndex(startIndex)
        val end = method.normalizeIndex(endIndex)
        val offset = if (method.hasIndex(endIndex)) 0 else 1 // this occurs on the end source
        this.instructions.addAll(instructions.subList(start, end + offset))
    }

    private fun constructExits(cfg: ControlFlowGraph): List<FlowBlock?> {
        return (0 until source.exits())
                .map { translate(cfg, source.exit(it)) }
                .toList()
    }

    private fun constructIncoming(cfg: ControlFlowGraph): List<FlowBlock?> {
        return (0 until source.incomings())
                .map { translate(cfg, source.incoming(it)) }
                .toList()
    }

    companion object {

        fun translate(cfg: ControlFlowGraph, block: ControlFlow.Block): FlowBlock? {
            return cfg.findBlockByIndex(block.position())
        }
    }
}
